### PR TITLE
Stable Context Menu Shortcuts

### LINKED ISSUE #896: [vim-plugin/rpc] Stable context menu short cuts
Currently the shortcuts for the context menu are [generated](https://github.com/phpactor/phpactor/blob/develop/autoload/phpactor/input.vim#L23) automatically based on the available choices.

The available choices change with every new feature, and so then do the shortcuts.

It would be better to define shortcuts in the RPC handler, or find another way of making them stable.

### PR DESCRIPTION
**Pull request by dantleech**
Allows the definition of key preferences to the context menu.

Note that we use VIM's `confirm` dialogue, which is case sensitive - which limits some our options when choosing keys.

Fixes #896

### CODE REVIEW

**Comment by dantleech** [comment]
cc @elythyr @przepompownia as you seem to have better VimL skills than me :)

File: autoload/phpactor/input.vim [hunk 1]
```diff
@@ -60,15 +54,56 @@ function! phpactor#input#choice(label, choices, ResultHandler)
     call a:ResultHandler(choices[choice - 1])
 endfunction
 
+function! s:determineConfirmLabelFromPreference(choiceLabel, preference)
+    let buffer = []
```
Comment (dantleech) [inline autoload/phpactor/input.vim]: not sure about variable scoping here -- should we use a scope prefix for local vars?

Comment (dantleech) [reply autoload/phpactor/input.vim]: they are local by default apparently :relieved:

Comment (przepompownia) [reply autoload/phpactor/input.vim]: Apparently (`:help l:` stands for edge cases but I cannot reproduce it, i.e. override `v:count` with local count), but it seems to be a good practice to explicitly define the scope (like method scope in PHP). If you use [vint](https://github.com/Vimjas/vint) (with [ALE](https://github.com/dense-analysis/ale) or [CoC](https://github.com/iamcco/diagnostic-languageserver/wiki/Linters) for example), you can notice that `ProhibitImplicitScopeBuiltinVariable` is disabled by default. This also does not look like common practice in most popular and advanced plugins.

File: autoload/phpactor/input.vim [hunk 2]
```diff
@@ -32,19 +33,12 @@ function! phpactor#input#choice(label, choices, ResultHandler)
...
-                call add(buffer, '&')
-                let foundShortcut = v:true
-                call add(usedShortcuts, tolower(char))
-            endif
-
-            call add(buffer, char)
-        endfor
-
-        let confirmLabel = join(buffer, "")
+        if !empty(a:keyMap) && has_key(a:keyMap, choiceLabel) && !empty(a:keyMap[choiceLabel])
```
Comment (przepompownia) [inline autoload/phpactor/input.vim]:
```suggestion
        if has_key(a:keyMap, choiceLabel) && !empty(a:keyMap[choiceLabel])
```
`v:false == has_key({}, 'x')` is true. No need to check if a:keyMap is empty.

File: autoload/phpactor/input.vim [hunk 3]
```diff
@@ -32,19 +33,12 @@ function! phpactor#input#choice(label, choices, ResultHandler)
...
+            let confirmLabel = s:determineConfirmLabelFromPreference(choiceLabel, a:keyMap[choiceLabel])
+        else
+            let confirmLabel = s:determineConfirmLabel(choiceLabel)
+        endif
 
         call add(list, confirmLabel)
         call add(choices, choiceLabel)
```
Comment (przepompownia) [inline autoload/phpactor/input.vim]: `choices` consist of all keys of `a:choises` and nothing else. It seems than you can simply assign `choices` to `keys(a:choices)` before this loop and iterate through it, isn't it?

File: autoload/phpactor/input.vim (hunk 2 above)
Comment (dantleech) [reply autoload/phpactor/input.vim]: `keyMap` can be a `list` when empty due to using an `array` as the JSON source :thinking:

File: autoload/phpactor/input.vim (hunk 3 above)
Comment (dantleech) [reply autoload/phpactor/input.vim]: indeed, well spotted.

File: autoload/phpactor/input.vim (hunk 1 above)
Comment (dantleech) [reply autoload/phpactor/input.vim]: Nice, I didn't know there was a static analysis tool for VimL :)

File: autoload/phpactor/input.vim (hunk 2 above)
Comment (przepompownia) [reply autoload/phpactor/input.vim]: Isn't `JSON_FORCE_OBJECT` helpful in such case?

Comment (dantleech) [reply autoload/phpactor/input.vim]: Well, we just switch to another problem then. There is probably a way to fix it on the PHP side, but I'm not that concerned.

Comment (przepompownia) [reply autoload/phpactor/input.vim]: Nothing too relevant to be concerned ;)

**Pull request merged**